xla/service/algebraic_simplifier.cc

```diff
@@ -8943,6 +8943,75 @@ Status AlgebraicSimplifierVisitor::HandleConvolution(
   if (swapped) {
     return OkStatus();
   }
+#if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
+  bool converted_conv = false;
+  bool can_rewrite = true;
+  auto from_dtype = convolution->shape().element_type();
+  if (options_.executing_on_cpu() && from_dtype == PrimitiveType::BF16) {
+    // Under the above conditions, convolutions remain in lower (BF16)
+    // precision. To ensure the correctness of the generated LLVM IR, we cast
+    // the convolutions to higher precision. This does not compromise
+    // performance as lower floating point precision convolutions are converted
+    // to higher precision in the regular optimization pipeline.
+    if (convolution->batch_group_count() != 1 ||
+        convolution->operand(1)->opcode() == HloOpcode::kReverse) {
+      can_rewrite = false;
+    }
+    const Shape& inp_shape = convolution->operand(0)->shape();
+    const Shape& ker_shape = convolution->operand(1)->shape();
+    const Shape& out_shape = convolution->shape();
+    if (ShapeUtil::IsZeroElementArray(inp_shape) ||
+        ShapeUtil::IsZeroElementArray(ker_shape) ||
+        ShapeUtil::IsZeroElementArray(out_shape)) {
+      can_rewrite = false;
+    }
+
+    auto dims = convolution->window().dimensions().size();
+    if (dims >= 4 || dims <= 0) can_rewrite = false;
+
+    if (inp_shape.rank() != ker_shape.rank() ||
+        inp_shape.rank() != out_shape.rank()) {
+      can_rewrite = false;
+    }
+
+    for (auto it = convolution->window().dimensions().begin();
+         it != convolution->window().dimensions().end(); it++) {
+      if ((*it).padding_low() < 0 || (*it).padding_high() < 0 ||
+          (*it).stride() < 0 || (*it).base_dilation() != 1 ||
+          (*it).window_reversal()) {
+        can_rewrite = false;
+      }
+    }
+
+    auto to_dtype = PrimitiveType::F32;
+
+    std::vector<HloInstruction*> new_operands;
+    auto from_dtype_operands = convolution->operands();
+
+    std::for_each(
+        from_dtype_operands.begin(), from_dtype_operands.end(),
+        [&new_operands, &to_dtype](HloInstruction* instr) {
+          new_operands.push_back(
+              instr->AddInstruction(HloInstruction::CreateConvert(
+                  ShapeUtil::ChangeElementType(instr->shape(), to_dtype),
+                  instr)));
+        });
+
+    HloInstruction* to_conv =
+        convolution->AddInstruction(convolution->CloneWithNewOperands(
+            ShapeUtil::ChangeElementType(convolution->shape(), to_dtype),
+            new_operands));
+
+    HloInstruction* from_conv =
+        to_conv->AddInstruction(HloInstruction::CreateConvert(
+            ShapeUtil::ChangeElementType(to_conv->shape(), from_dtype),
+            to_conv));
+
+    TF_RETURN_IF_ERROR(ReplaceInstruction(convolution, from_conv));
+    convolution = to_conv;
+    converted_conv = true;
+  }
+#endif  // INTEL_MKL && ENABLE_ONEDNN_V3
```

Comment:
Yes, the FloatSupport handles the conversions. However, because Algebraic Simplifier precedes the oneDNN rewriter, there can be instances where the Algebraic Simplifier may transform oneDNN rewritable BF16 convolutions into BF16 Muls and Dots which can create problems.
This can be removed once BF16 Muls and Dots are fully supported.